refactor(compiler): integrate let declarations into the template pipeline

crisbeto opened the pull request:
These changes integrate let declarations into the template pipeline. This involves a few operations:
* Producing a `declareLet` instruction call at creation time to initialize the declaration.
* Producing a `storeLet` instruction call in the place of the let declaration, including the necessary `advance` calls beforehand.
* For let declarations used within their declaration view, moving the `const` to be placed right after the `storeLet` call to ensure the their value has been computed.
* For let declarations that are _only_ used in their declaration view, removing the `storeLet` call and inlining the expression into the constant statement.

crisbeto commented on packages/compiler/src/template/pipeline/src/phases/generate_variables.ts:
```diff
@@ -190,6 +231,7 @@ function getScopeForView(view: ViewCompilationUnit, parent: Scope | null): Scope
 function generateVariablesInScopeForView(
   view: ViewCompilationUnit,
   scope: Scope,
+  isListener: boolean,
```
I don't feel great about adding a parameter like this, but I couldn't find another way to distinguish between listener and non-listener scopes.

alxhub replied:
I'm curious why we need to? Naively, I'd expect that we _should_ generate the same variables inside listener functions. Or am I missing something?

crisbeto replied:
It's because, while in the current setup of the pipeline listeners have the same scope as their declaration view, they're technically child views, hence why we resolve `@let` declarations in them using `ɵɵreadContextLet` and why they currently generate `nextContext` calls.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/reify.ts:
```diff
@@ -537,6 +540,20 @@ function reifyUpdateOperations(_unit: CompilationUnit, ops: ir.OpList<ir.UpdateO
+        ir.OpList.replace<ir.UpdateOp>(
+          op,
+          ir.createStatementOp(
+            op.isUsedAcrossViewBoundaries
```
Logic like this shouldn't go in the reify phase. Think of reify as a direct translation from the IR to TS code - it shouldn't be making any decisions about the semantics of the code we generate.

The template pipeline has a variable system that represents declared variables. Ideally, there would be a phase which translates a `LetOp` to a `VariableOp` with a `StoreLet` expression initializer. Then, the variable optimization phase should be able to remove that variable if it's not used/required.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/resolve_names.ts:
```diff
@@ -34,7 +34,7 @@ function processLexicalScope(
   // Since variables are generated in each view for the entire lexical scope (including any
   // identifiers from parent templates) only local variables need be considered here.
-  const scope = new Map<string, ir.XrefId>();
+  const scope = new Map<string, ir.VariableOp<ir.CreateOp | ir.UpdateOp> | ir.StoreLetOp>();
```
This also feels like a separation of concerns issue. The resolve_names phase shouldn't have any special awareness of `@let`.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/resolve_names.ts:
```diff
@@ -120,6 +139,28 @@ function processLexicalScope(
+ * @param expressionOp Op that is referring to a `@let` declaration.
+ * @param target Op the declares the `@let`.
+ */
+function isIllegalForwardLetReference(
```
This feels like something a dedicated phase should do.

crisbeto committed:
refactor(compiler): integrate let declarations into the template pipeline

These changes integrate let declarations into the template pipeline. This involves a few operations:
* Producing a `declareLet` instruction call at creation time to initialize the declaration.
* Producing a `storeLet` instruction call in the place of the let declaration, including the necessary `advance` calls beforehand.
* For let declarations used within their declaration view, moving the `const` to be placed right after the `storeLet` call to ensure the their value has been computed.
* For let declarations that are _only_ used in their declaration view, removing the `storeLet` call and inlining the expression into the constant statement.

crisbeto committed:
fixup! refactor(compiler): integrate let declarations into the template pipeline

crisbeto commented on packages/compiler/src/template/pipeline/src/phases/variable_optimization.ts:
```diff
@@ -48,6 +48,10 @@ export function optimizeVariables(job: CompilationJob): void {
+
+  // `@let` declarations can only be optimized once the rest of the variables were
+  // optimized. Run it as a separate pass after the main optimization pass.
+  optimizeLetVariables(job);
```
I was a bit on the fence whether to do this here or in a separate phase. I did it here, because:
1. It has to run _after_ all other variables have been optimized.
2. It is optimizing variables.

dylhunn reviewed (commented):
As per VC -- my only suggested structural change is that optimization for let should probably be its own phase, since it has let-specific knowledge (e.g. extracting initializers from `storeLet`).

Thanks for the nice work!

crisbeto committed:
fixup! refactor(compiler): integrate let declarations into the template pipeline

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_let/let_multiple_optimization_template.js:
```diff
@@ -0,0 +1,18 @@
+  template: function MyApp_Template(rf, ctx) {
+    if (rf & 1) {
+      $r3$.ɵɵtext(0);
+      $r3$.ɵɵdeclareLet(1)(2)(3)(4);
```
Since the reads are optimized out, can we do the same for the declaration and therefore the reserved binding slots?

crisbeto replied:
I was considering this as well, but decided not to do it for now after chatting with Pawel. I left the context here: https://github.com/angular/angular/pull/56299/files#diff-f271e852d3ef9ded1fc03c94fead7cdc6e6c7d645202ad33399d2f7d63a0a9a4R24-R29

crisbeto commented on packages/compiler/src/template/pipeline/src/phases/resolve_names.ts:
```diff
@@ -61,6 +61,9 @@ function processLexicalScope(
         }
         break;
+      case ir.OpKind.StoreLet:
+        scope.set(op.declaredName, op.target);
```
I'll double-check it, it might be a leftover from one of the earlier iterations.

crisbeto replied:
Removed.

crisbeto commented on packages/compiler/src/template/pipeline/src/phases/variable_optimization.ts:
```diff
@@ -178,6 +178,12 @@ function optimizeVariablesInOpList(
     const opInfo = opMap.get(op)!;
 
     if (op.kind === ir.OpKind.Variable && varUsages.get(op.xref)! === 0) {
+      // `storeLet` variables are optimized in a separate pass once
```
I tried a few times to make them work through the current optimization phase, but arrived to the conclusion that we can't. The problem is that variable optimization needs to happen so that we know _how_ the `@let`-based variables are used and we can transform them accordingly. I listed out the 3 possibilities for the generated code here: https://github.com/angular/angular/pull/56299/files#diff-f271e852d3ef9ded1fc03c94fead7cdc6e6c7d645202ad33399d2f7d63a0a9a4R14. Specifically for the case where a variable is only used within the same view we first have to eliminate all the usages in nested views.

alxhub replied:
My suggestion is that we do both:

* first run them through the normal variable optimizer
* then have a second pass which is focused on optimizing the `storeLet` call

So, the variable optimizer could do its normal thing, and would convert `let x = storeLet(...);` to `storeLet(...);` if the local variable is unused. It could also choose to inline the `storeLet` call into a usage site (maybe?) if only used once.

Then, a dedicated `storeLet` optimizer would come along and perform one of two optimizations:

* in the case of `storeLet(expr);`, if the `let` variable is never read outside of the local template, it could be unwrapped into a plain `expr`.
* if this results in an expression statement of `expr;`, it can be removed entirely (maybe?)

crisbeto commented on packages/compiler/src/template/pipeline/src/phases/generate_variables.ts:
```diff
@@ -61,14 +61,32 @@ function recursivelyProcessView(view: ViewCompilationUnit, parentScope: Scope |
-  // Prepend the declarations for all available variables in scope to the `update` block.
-  const preambleOps = generateVariablesInScopeForView(view, scope);
-  view.update.prepend(preambleOps);
+  for (const op of view.update) {
```
Do you mean specifically the transformation of `StoreLet` to `VariableOp` or prepending the variables referencing `StoreLet` from parent views to child views? I can see a case for the former, but I think the latter should stay in the current phase.

crisbeto replied:
Moved the logic for replacing local let references into a separate phase.

crisbeto committed:
fixup! refactor(compiler): integrate let declarations into the template pipeline

crisbeto committed:
fixup! refactor(compiler): integrate let declarations into the template pipeline

crisbeto committed:
fixup! refactor(compiler): integrate let declarations into the template pipeline

alxhub commented on packages/compiler/src/template/pipeline/src/phases/variable_optimization.ts:
```diff
@@ -320,7 +320,10 @@ function fencesForIrExpression(expr: ir.Expression): Fence {
     case ir.ExpressionKind.RestoreView:
       return Fence.ViewContextRead | Fence.ViewContextWrite | Fence.SideEffectful;
+    case ir.ExpressionKind.StoreLet:
+      return Fence.SideEffectful | Fence.ViewContextWrite;
```
Is `StoreLet` actually `ViewContextWrite`? It doesn't change the current view context (the name is a bit confusing - `ViewContextWrite` means it changes the current view context in the runtime.

I'm not even sure it's `ViewContextRead` - it doesn't actually depend on the view context at all.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/store_let_optimization.ts:
```diff
@@ -0,0 +1,50 @@
+  // TODO(crisbeto): potentially remove the unused calls completely, pending discussion.
+  for (const unit of job.units) {
+    for (const op of unit.update) {
+      if (op.kind === ir.OpKind.StoreLet && !letUsedExternally.has(op.target)) {
```
`StoreLet` ops should be gone by now - long ago transformed into a `StoreLetExpr`.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/reify.ts:
```diff
@@ -537,6 +540,12 @@ function reifyUpdateOperations(_unit: CompilationUnit, ops: ir.OpList<ir.UpdateO
         ir.OpList.replace(op, ng.deferWhen(op.prefetch, op.expr, op.sourceSpan));
         break;
+      case ir.OpKind.StoreLet:
+        ir.OpList.replace<ir.UpdateOp>(
```
`StoreLet` ops should be completely gone - we should error if we find one.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/generate_local_let_references.ts:
```diff
@@ -0,0 +1,39 @@
+export function generateLocalLetReferences(job: ComponentCompilationJob): void {
+  for (const unit of job.units) {
+    for (const op of unit.update) {
+      if (op.kind === ir.OpKind.StoreLet) {
```
Prefer early continue:
```typescript
if (!op.kind === ir.OpKind.StoreLet) {
  continue;
}

// main logic is not indented.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/generate_advance.ts:
```diff
@@ -36,16 +37,33 @@ export function generateAdvance(job: CompilationJob): void {
-      if (!ir.hasDependsOnSlotContextTrait(op)) {
+      let consumer: ir.DependsOnSlotContextOpTrait;
+
+      if (ir.hasDependsOnSlotContextTrait(op)) {
```
Rather than hardcoding specific forms to check here, we should use a visitor to check all expressions in the op for `ir.hasDependsOnSlotContextTrait`.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/remove_illegal_let_references.ts:
```diff
@@ -0,0 +1,46 @@
+export function removeIllegalLetReferences(job: CompilationJob): void {
+  for (const unit of job.units) {
+    for (const op of unit.update) {
+      if (
```
Switch to early continue pattern to avoid indentation.

crisbeto committed:
fixup! refactor(compiler): integrate let declarations into the template pipeline
